### PR TITLE
honor CPE logical expressions when generating Bash and Ansible remediations

### PR DESCRIPTION
**Pull request by vojtapolasek**
What currently does not work is:
- when generating Ansible conditionals, platform_package_overrides is not honored. No idea why, I need help.
- one unit SSG test needs to be modified.
Other things should work very nicely, I did comparisons of datastreams generated from this branch and from master and we should not break anything.

#### Description:

- create new Jinja macros for creating of conditionals for installed packages
- put correct conditionals to definitions of CPE items (shared/applicability/...)
- get ridf of some magic functions in build_remediations.py (platform package overrides, selection of packaging system) this is now handled by JInja
- some tweaks to parsing of conditionals, especially ignore them if they are empty

#### Rationale:

- _Rationale here. Replace this text. Don't use the italics format!_

- Fixes # _Issue number here (e.g. #26) or remove this line if no issue exists._

### CODE REVIEW

**Comment by jan-cerny** [comment]
@vojtapolasek Regarding the Ansible:

> when generating Ansible conditionals, platform_package_overrides is not honored. No idea why, I need help.

It seems to work for me, for example in the `debian10` product which has multiple `platform_packaged_overrides`, the compiled platforms contain correct `ansible_conditional`. Except the `login_defs` platform, when I found that there is a typo in `shared/applicability/general.yml`: on line 35 there needs to be an underscore (`_`) in `login_defs` instead of a dash.

File: shared/applicability/general.yml [hunk 1]
```diff
@@ -3,41 +3,57 @@ cpes:
...
       check_id: installed_env_has_libuser_package
+      bash_conditional: {{{ bash_pkg_conditional("libuser") }}}
+      ansible_conditional: {{{ ansible_pkg_conditional("libuser") }}}
 
   - login_defs:
       name: "cpe:/a:login_defs"
       title: "Package providing /etc/login.defs is installed"
       check_id: installed_env_has_login_defs
+      bash_conditional: {{{ bash_pkg_conditional("login_defs") }}}
+      ansible_conditional: {{{ ansible_pkg_conditional("login-defs") }}}
```
Comment (jan-cerny) [inline shared/applicability/general.yml:35]:
```suggestion
      ansible_conditional: {{{ ansible_pkg_conditional("login_defs") }}}
```

File: ssg/build_cpe.py [hunk 2]
```diff
@@ -153,6 +153,12 @@ def __init__(self, cpeitem_data):
         self.name = cpeitem_data["name"]
         self.title = cpeitem_data["title"]
         self.check_id = cpeitem_data["check_id"]
+        self.bash_conditional = ""
+        if "bash_conditional" in cpeitem_data.keys():
+            self.bash_conditional = cpeitem_data["bash_conditional"]
+        self.ansible_conditional = ""
+        if "ansible_conditional" in cpeitem_data.keys():
+            self.ansible_conditional = cpeitem_data["ansible_conditional"]
```
Comment (jan-cerny) [inline ssg/build_cpe.py:161]: If  you look for a key in a dictionary, you don't have to use `x in d.keys()` it's just `x in d`. Also, you can use `.get()` to which gives a default value if a key isn't present in the dictionary.

```suggestion
        self.bash_conditional = cpeitem_data.get("bash_conditional", "")
        self.ansible_conditional = cpeitem_data.get("ansible_conditional", "")
```

**Review by jan-cerny — APPROVED** [review]

**Comment by jan-cerny** [comment]
@vojtapolasek Can I merge now? Also, you mentioned in this PR description that one unit SSG test needs to be modified. Which test is that?

**Comment by vojtapolasek** [comment]
You can merge if it helps you in writing test.
The failing test is python-unit-ssg-module .

**Comment by vojtapolasek** [comment]
I have removed it.

**Comment by yuumasato** [comment]
Removing milestone again, :)

**Pull request merged**